x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkValidateUtil.java

```diff
@@ -0,0 +1,375 @@
+            );
+        } else if (Strings.hasText(jwkSetContentsHmac) && (allowedSignatureAlgorithmsHmac.isEmpty())) {
+            // HMAC JWKSet present without any HMAC signature algorithms
+            throw new SettingsException(
+                "Settings mismatch. HMAC JWKSet is set in ["
+                    + jwkSetConfigKeyHmac
+                    + "], but no HMAC signature algorithms are set in ["
+                    + allowedSignatureAlgorithmsConfigKey
+                    + "]"
+            );
+        } else if ((allowedSignatureAlgorithmsHmac.isEmpty() == false) && (Strings.hasText(jwkSetContentsHmac) == false)) {
+            // HMAC signature algorithms present without a HMAC JWKSet
+            throw new SettingsException(
+                "Settings mismatch. HMAC signature algorithms are set in ["
+                    + allowedSignatureAlgorithmsConfigKey
+                    + "], but no HMAC JWKSet is set in ["
+                    + jwkSetConfigKeyHmac
+                    + "]"
+            );
+        } else if (Strings.hasText(jwkSetContentsPkc) && (allowedSignatureAlgorithmsPkc.isEmpty())) {
+            // RSA/EC JWKSet present without any RSA/EC signature algorithms
+            throw new SettingsException(
+                "Settings mismatch. RSA/EC JWKSet is set in ["
+                    + jwkSetConfigKeyPkc
+                    + "], but no RSA/EC signature algorithms are set in ["
+                    + allowedSignatureAlgorithmsConfigKey
+                    + "]"
+            );
+        } else if ((allowedSignatureAlgorithmsPkc.isEmpty() == false) && (Strings.hasText(jwkSetContentsPkc) == false)) {
+            // RSA/EC signature algorithms present without an RSA/EC JWKSet
+            throw new SettingsException(
+                "Settings mismatch. RSA/EC signature algorithms are set in ["
+                    + allowedSignatureAlgorithmsConfigKey
+                    + "], but no RSA/EC JWKSet is set in ["
+                    + jwkSetConfigKeyPkc
+                    + "]"
+            );
+        }
+
+        // At this point, one or both JWKSets are configured. Load and validate their JWK contents against each set of signature algorithms.
+        JWKSet jwkSetHmac = null;
+        if (Strings.hasText(jwkSetContentsHmac)) {
+            jwkSetHmac = JwkValidateUtil.loadJwkSetFromString(jwkSetConfigKeyHmac, jwkSetContentsHmac);
+            JwkValidateUtil.validateJwkSetContents(jwkSetConfigKeyHmac, jwkSetHmac, allowedSignatureAlgorithmsHmac);
+        }
+        JWKSet jwkSetPkc = null;
+        if (Strings.hasText(jwkSetContentsPkc)) {
+            jwkSetPkc = JwkValidateUtil.loadJwkSetFromString(jwkSetConfigKeyPkc, jwkSetContentsPkc);
+            JwkValidateUtil.validateJwkSetContents(jwkSetConfigKeyPkc, jwkSetPkc, allowedSignatureAlgorithmsPkc);
+        }
+        return new Tuple<>(jwkSetHmac, jwkSetPkc);
+    }
+
+    public static void validateJwkSetContents(
```

Comment:
I will echo my comments from previous iterations: I'm not sure it's worth the complexity/maintainance costs to add these checks. I won't insist on removing but I have two asks:

1. When we implement reloading of JWKSets ( from URL or file), we need to make sure we do the same checks again
2. I think we should not be failing in all these cases but instead WARN log, _as long as_ we have 1 usable JWK after completing the validation. For instance:
   a. Why should we fail to load the  JWKSets and use it just because one of 99 keys in it has the  wrong bit length, wrong algorithm ? The rest 98 are perfectly fine. 
   b. Why should we fail to load the realm if the JWKSet contain an extra key which is not of a type (not RSA, EC , HMAC) that we can use ? As long there is one we can use, we should be ok
   c. Why should we fail if there is an RSA key in the JWKset but the _current_ configuration only wants EC ? The OP might support all kinds of algorithms for different RPs so it's common that they have keys in the JWKSet for all algorithms. We shouldn't fail to start the realm if we want to only use RSA ourselves.

Reply:
Refactored. Changed failures to warnings. Algs and JWKs are filtered.
- Any alg without corresponding JWK is ignored, and logged.
- Any JWK without corresponding alg is ignored.

If one usable alg and JWK are left over, the realm will continue.

Moved PKC code to a reusable method. If update drops the usable alg or JWK count to zero, update is ignored and the previous settings continue to be used.

Reply:
> Moved PKC code to a reusable method. If update drops the usable alg or JWK count to zero, update is ignored and the previous settings continue to be used.

I don' think this is what we want. The key update happens for a number of reasons, rotation, compromise etc. It would be wrong to keep using old/compromised keys because we can't parse the new key set. It sucks a little on runtime because we can't fail to start the node, but I think the correct behavior in this case it to keep an empty list of JWKs in memory and fail all authentications with a relevant error and printing out a proper warning that the administrator can easily see in the logs